### PR TITLE
support rust-analyzer macro expansion

### PR DESCRIPTION
**Pull request by brotzeit**

### CODE REVIEW

File: lsp-rust.el [hunk 1]
```diff
@@ -335,6 +335,11 @@ PARAMS progress report notification data."
   :type 'boolean
   :package-version '(lsp-mode . "6.2"))
 
+(defcustom lsp-rust-analyzer-macro-expansion-method 'lsp-rust-analyzer-macro-expansion-default
+  "Use a different function if you want formatted macro expansion results and syntax highlighting."
```
Comment (brotzeit) [inline lsp-rust.el]: I will add functions in rust-mode and rustic.

File: lsp-rust.el [hunk 2]
```diff
@@ -484,6 +489,32 @@ PARAMS progress report notification data."
...
                                      (lsp-rust-analyzer-inlay-hints-mode)))))
 
+(defun lsp-rust-analyzer-expand-macro ()
+  "Expands the macro call at point recursively."
+  (interactive)
+  (-if-let ((workspace (lsp-find-workspace 'rust-analyzer default-directory)))
+      (-if-let* ((params (list :textDocument (lsp--text-document-identifier)
+                               :position (lsp--cur-position)))
+                 (response (with-lsp-workspace workspace
+                             (lsp-send-request (lsp-make-request
```
Comment (yyoncho) [inline lsp-rust.el]: nit: you may use lsp-request and lsp--text-document-position-params

File: lsp-rust.el [hunk 3]
```diff
@@ -484,6 +489,32 @@ PARAMS progress report notification data."
...
+                                                "rust-analyzer/expandMacro"
+                                                params))))
+                 (result (ht-get response "expansion")))
+          (funcall lsp-rust-analyzer-macro-expansion-method result)
+        (message "No macro found at point, or it could not be expanded."))
```
Comment (yyoncho) [inline lsp-rust.el]: nit: you may use lsp--error.

**Comment by brotzeit** [comment]
> Is it possible to expand the macros inline like some of the lisp modes do?

I've never seen this for macros. Which packages provide this feature ?

**Comment by yyoncho** [comment]
> > Is it possible to expand the macros inline like some of the lisp modes do?
> 
> I've never seen this for macros. Which packages provide this feature ?

https://github.com/joddie/macrostep (cider do that as well).

File: lsp-rust.el (hunk 2 above)
Comment (brotzeit) [reply lsp-rust.el]: I will make these changes for the whole file in a new pull request.

**Pull request merged**